### ISSUE TITLE
Refs to other localhosts do not work

### ISSUE THREAD

**Issue by DB-Alex** [storybookjs/storybook#18405]
Bug:

Access to fetch at 'http://localhost:3000/stories.json' from origin 'http://localhost:3020' has been blocked by CORS policy: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'.

You can not use the wildcard * when the credentials mode is true. At the moment it is not set so I think chrome is placing:

Access-Control-Allow-Credentials: true

by default. I think when this is set to:

Access-Control-Allow-Credentials: false

it should work.

**Comment by Mrfaann** [comment]
yes, it should be worked[.](https://basketballguider.com/) Can you share more detail for more understanding.!

**Comment by burakkgunduzz** [comment]
facing with the same problem, how can I solve it any thoughts?

**Comment by itsik-avidan-gong** [comment]
I'm encountering the same issue. 
I have a dev server without any stories with `ref` pointing to 2 other dev servers on my localhost, all are configures with `buildStoriesJson: true`. It seems that there is a race condition on the server start between the hosting server to the child servers, resulting with CORS error when the host is trying to access `/stories.json` path. 

env details : 
node v16.13.1
storybook v6.5.10
pnpm v7.9.0 ( as scripts executer )


I've managed to create a workaround by adding a wait in my load order, but clearly this not the ideal solution

```
"child1:storybook": "start-storybook -p 6006 -c \"./child1/.storybook\"",
"child2:storybook": "start-storybook -p 6007 -c \"./child2/.storybook\"",
"paret:storybook": "sleep 5 && start-storybook -p 6001 -c \"./.storybook\"",
"storybook-all": "npm-run-all -p child1:storybook child2:storybook parent:storybook -l",
```

also according to https://github.com/storybookjs/storybook/issues/14254 it seems to be a regression

**Comment by EloB** [comment]
Also face this issue. One main storybook that consume 2 other storybook.

**Comment by dylannnn** [comment]
Same issue on version 7.0.0-beta.33

**Comment by hdwatts** [comment]
Same issue on version 7.0.0-beta.35

**Comment by penx** [comment]
Confirmed in version 6.5.14, if I start the parent storybook first and then the child I get this error.

Starting the child storybook first, followed by the parent, resolves this error for me.

**Comment by kaydeejay** [comment]
> Confirmed in version 6.5.14, if I start the parent storybook first and then the child I get this error.
> 
> Starting the child storybook first, followed by the parent, resolves this error for me.

That resolution worked for me before version 7, but since upgrading that no longer works for me. I've tried using concurrently and wait-on, but I'm unable to successfully run more than 1 storybook with concurrently. One of them will always give 404 errors. They both work when I run them independently.

**Comment by doworld-dev** [comment]
> 흠, 같은 오류가 발생했습니다. 문제는 모든 항목을 무작위 순서로 한 번에 회전시키는 Turborepo `--parallel`를 사용할 때 발생합니다.
> 
> 먼저 하위 항목을 수동으로 시작한 다음 구성 구성이 있는 루트/상위 프로젝트를 시작하면 모든 것이 잘 작동합니다.
> 
> 무엇을 먼저 시작하든 상관없다면 정말 좋을 것 같습니다.

I'm having the same problem.
Did you solve it?

**Comment by is-paranoia** [comment]
Have same problem. 
I'm trying to compose sb8 and sb6
Does anyone solve it? How?

**Comment by pstrh** [comment]
Same issue here on v8.2.9.

The issue seems to be related to how storybook sets the credentials property of the fetch API. 

See: 
- https://github.com/storybookjs/storybook/blob/221f6a78332adb67450ad94f172aac6cf490de12/code/core/src/manager-api/modules/refs.ts#L214 
- https://github.com/storybookjs/storybook/blob/221f6a78332adb67450ad94f172aac6cf490de12/code/core/src/common/utils/get-storybook-refs.ts#L122

In my case storybook determines that my storybook to be embedded is not public and sets credentials to "include" - but actually it doesn't need any credentials to be called and is publicly available.  

BTW: If I change the code (temporarily in node_modules) from "include" to "omit" it loads the index.json and the stories.json. But then I run into another issues saying that metadata.json is missing. I had to manually add a metadata.json with the following content to fix this. 

metadata.json:
```
{
    "versions": {}
}
``` 
Only with these two fixes I got the composition feature running locally.

**Comment by pstrh** [comment]
As a workaround to fix the CORS issue with localhost I've added a proxy to the vite config in the main.ts/.js:

```
const config: StorybookConfig = {
  ...,
  framework: {
    name: "@storybook/react-vite",
    options: {}
  },
  refs: {
    designsystem: {
      title: "My design system",
      url: "/designsystem",
    }
  },
  async viteFinal(config) {
    return mergeConfig(config, defineConfig({
      server: {
        "/designsystem": {
          target: "http://localhost:3000",
          rewrite: (path: string) => path.replace(/^\/designsystem/, "")
        }
      }
    }));
  }
};
```
I also found that the composition works if the requests to stories.json and metadata.json fail. The requests are done for compatibility reasons to support older storybook versions or for advanced use cases. Therefore, the metadata.json mentioned above is not needed.

**Comment by LeonxLJX** [comment]
I'd like to take this one.

I've read through the report. Before changing anything I'd like to reproduce it against current `main` first — my reasoning is that a fix anchored to the actual observed behavior is much easier to review than one inferred from the report alone, especially if the surrounding code has moved since this was filed.

Plan: reproduce and confirm the root cause, then make the smallest change that addresses it (plus a regression test if the repo has coverage for this area). If the root cause turns out to live somewhere other than where the report points, I'll report back here before forcing a change.

Anything already known about this that I should read first?

(claiming via @LeonxLJX)
